chat-android/src/main/java/com/ably/chat/AtomicCoroutineScope.kt

```diff
@@ -0,0 +1,82 @@
+package com.ably.chat
+
+import io.ably.annotation.Experimental
+import java.util.concurrent.PriorityBlockingQueue
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+
+/**
+ * AtomicCoroutineScope is a thread safe wrapper to run multiple operations mutually exclusive.
+ * All operations are atomic and run with given priority.
+ * Accepts scope as a constructor parameter to run operations under the given scope.
+ * See [Kotlin Dispatchers](https://kt.academy/article/cc-dispatchers) for more information.
+ */
+class AtomicCoroutineScope(private val scope: CoroutineScope = CoroutineScope(Dispatchers.Default)) {
+
+    private val sequentialScope: CoroutineScope = CoroutineScope(Dispatchers.Default.limitedParallelism(1))
+
+    private class Job(
```

Comment:
To be honest, Job is not a very good name for this class, because in the Coroutine context, Job is heavily used.

Reply:
Yeah, true. But class is private and not exposed externally. Also I thought, in the context of `AtomicCoroutineScope` it made sense to keep the name as `Job`.